### PR TITLE
config-tools: board_info.h and pci_dev.c xfom using xslt

### PR DESCRIPTION
**Pull request by yuchuyang**
Generate board_info.h and pci_dev.c based on the new board.xml and integrated to acrn project compilation.

Two static allocators are added: mmio.py and bdf.py. Those allocate free mmio windows and bdf to emulated vuarts, inter-vm shared memory and passthrough devices

### CODE REVIEW

**Comment by yuchuyang** [comment]
This PR need to work with the new board.xml which is generated using https://github.com/projectacrn/acrn-hypervisor/pull/6008.

**Comment by wenlingz** [comment]
@yuchuyang  build fail, please refer:

ERROR: acrn-hypervisor-native-6026-1621322514 do_compile: oe_runmake failed
ERROR: acrn-hypervisor-native-6026-1621322514 do_compile: Execution of '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621322514/temp/run.do_compile.903' failed with exit code 1:
make: Entering directory '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621322514/git/hypervisor'
Board XML is fetched from /home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621322514/git/misc/config_tools/data/ehl-crb-b/ehl-crb-b.xml
Scneario XML is configuration fetched from /home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621322514/git/misc/config_tools/data/ehl-crb-b/hybrid_rt.xml
Traceback (most recent call last):
  File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621322514/git/misc/config_tools/static_allocators/main.py", line 41, in <module>
    main(args)
  File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621322514/git/misc/config_tools/static_allocators/main.py", line 31, in main
    module.fn(board_etree, scenario_etree, allocation_etree)
  File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621322514/git/misc/config_tools/static_allocators/gpa.py", line 282, in fn
    create_native_pci_hole_node(allocation_etree, native_low_mem, native_high_mem)
  File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621322514/git/misc/config_tools/static_allocators/gpa.py", line 239, in create_native_pci_hole_node
    common.append_node("/acrn-config/hv/MMIO/LOW_MMIO_START", hex(low_mem[0].start).upper(), allocation_etree)
IndexError: list index out of range
make: *** No rule to make target '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621322514/build//hypervisor/configs/allocation.xml', needed by '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621322514/build//hypervisor/configs/unified.xml'.  Stop.
make: Leaving directory '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621322514/git/hypervisor'
WARNING: /home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621322514/temp/run.do_compile.903:1 exit 1 from 'exit 1'

ERROR: Logfile of failure stored in: /home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621322514/temp/log.do_compile.903
Log data follows:
| DEBUG: Executing shell function do_compile

**Comment by wenlingz** [comment]
@yuchuyang please rebase your code, thanks

File: misc/config_tools/xforms/board_info.h.xsl [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  <xsl:template name="MAX_VMSIX_ON_MSI_PDEVS_NUM">
+    <xsl:value-of select="acrn:define('MAX_VMSIX_ON_MSI_PDEVS_NUM', sum(dyn:map(//device, 'acrn:is-vmsix-supported-device(./vendor, ./identifier)')), 'U')" />
+  </xsl:template>
+
+  <xsl:template match="MMIO">
+    <xsl:value-of select="acrn:define('MMIO32_START', //MMIO32_START, 'UL')" />
+    <xsl:value-of select="acrn:define('MMIO32_END', //MMIO32_END, 'UL')" />
+    <xsl:value-of select="acrn:define('MMIO64_START', //MMIO64_START, 'UL')" />
+    <xsl:value-of select="acrn:define('MMIO64_END', //MMIO64_END, 'UL')" />
+    <xsl:value-of select="acrn:define('HI_MMIO_START', //HI_MMIO_START, 'UL')" />
```
Comment (lifeix) [inline misc/config_tools/xforms/board_info.h.xsl:72]: Why "HI_MMIO_START" and "HI_MMIO_END" still exist ? I think you have already replaced them with MMIO64_START and MMIO64_END.

Comment (yuchuyang) [reply misc/config_tools/xforms/board_info.h.xsl:72]: They are still in used, and have different generated rule.

**Comment by yuchuyang** [comment]
@wenlingz  This patch is using the xpath which can only work with the newer board.xml... do we have plan to update the CI board.xmls?

**Comment by wenlingz** [comment]
> @wenlingz This patch is using the xpath which can only work with the newer board.xml... do we have plan to update the CI board.xmls?

If there have PR update the new XML, CI will use the latest xml

**Comment by wenlingz** [comment]
OK to verify

**Comment by wenlingz** [comment]
@yuchuyang build fail whl -industry

ERROR: acrn-hypervisor-native-6026-1621565902 do_compile: oe_runmake failed
ERROR: acrn-hypervisor-native-6026-1621565902 do_compile: Execution of '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621565902/temp/run.do_compile.738' failed with exit code 1:
make: Entering directory '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621565902/git/hypervisor'
Board XML is fetched from /home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621565902/git/misc/config_tools/data/whl-ipc-i5/whl-ipc-i5.xml
Scneario XML is configuration fetched from /home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621565902/git/misc/config_tools/data/whl-ipc-i5/industry.xml
[1;33mWarning[0m: All available pio bases are used by native fully. '0x2F8' is taken by sos legacy vuart 1.
Traceback (most recent call last):
  File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621565902/git/misc/config_tools/static_allocators/main.py", line 41, in <module>
    main(args)
  File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621565902/git/misc/config_tools/static_allocators/main.py", line 31, in main
    module.fn(board_etree, scenario_etree, allocation_etree)
  File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621565902/git/misc/config_tools/static_allocators/gpa.py", line 305, in fn
    create_native_pci_hole_node(allocation_etree, native_low_mem, native_high_mem)
  File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621565902/git/misc/config_tools/static_allocators/gpa.py", line 238, in create_native_pci_hole_node
    common.append_node("/acrn-config/hv/MMIO/MMIO32_START", hex(low_mem[0].start).upper(), allocation_etree)
IndexError: list index out of range
make: *** No rule to make target '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621565902/build//hypervisor/configs/allocation.xml', needed by '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621565902/build//hypervisor/configs/unified.xml'.  Stop.
make: Leaving directory '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621565902/git/hypervisor'
WARNING: /home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621565902/temp/run.do_compile.738:1 exit 1 from 'exit 1'

ERROR: Logfile of failure stored in: /home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621565902/temp/log.do_compile.738
Log data follows:
| DEBUG: Executing shell function do_compile
| NOTE: make RELEASE=0 SYSROOT= O=/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621565902/build/ DESTDIR=/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux

**Comment by yuchuyang** [comment]
@wenlingz let me know if this still fails the test, otherwise it should be good to go.

**Comment by wenlingz** [comment]
> @wenlingz let me know if this still fails the test, otherwise it should be good to go.

OK, will feedback

**Comment by wenlingz** [comment]
@yuchuyang still build fail, please refer:

acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621582614/build//hypervisor BOARD=whl-ipc-i5 SCENARIO=industry -C hypervisor pre_build
| make: Entering directory '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621582614/git/hypervisor'
| Board XML is fetched from /home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621582614/git/misc/config_tools/data/whl-ipc-i5/whl-ipc-i5.xml
| Scneario XML is configuration fetched from /home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621582614/git/misc/config_tools/data/whl-ipc-i5/industry.xml
| Traceback (most recent call last):
|   File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621582614/git/misc/config_tools/static_allocators/main.py", line 41, in <module>
|     main(args)
|   File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621582614/git/misc/config_tools/static_allocators/main.py", line 31, in main
|     module.fn(board_etree, scenario_etree, allocation_etree)
|   File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621582614/git/misc/config_tools/static_allocators/gpa.py", line 305, in fn
|     create_native_pci_hole_node(allocation_etree, native_low_mem, native_high_mem)
|   File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621582614/git/misc/config_tools/static_allocators/gpa.py", line 238, in create_native_pci_hole_node
|     common.append_node("/acrn-config/hv/MMIO/MMIO32_START", hex(low_mem[0].start).upper(), allocation_etree)
| IndexError: list index out of range
| make: *** No rule to make target '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621582614/build//hypervisor/configs/allocation.xml', needed by '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621582614/build//hypervisor/configs/unified.xml'.  Stop.
| make: Leaving directory '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621582614/git/hypervisor'
| ERROR: oe_runmake failed
| WARNING: /home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621582614/temp/run.do_compile.487:1 exit 1 from 'exit 1'
| ERROR: Execution of '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621582614/temp/run.do_compile.487' failed with exit code 1:
| make: Entering directory '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6026-1621582614/git/hypervisor'

**Comment by yuchuyang** [comment]
@wenlingz looks like CI tests whl-i5-ipc board which is not using new board xml...
Do we have a test suites list of CI?

**Comment by wenlingz** [comment]
OK to verify

**Comment by wenlingz** [comment]
try verify EHL

**Review by NanlinXie — APPROVED** [review]

**Pull request merged**